core/src/ledger_cleanup_service.rs

```diff
@@ -62,37 +65,88 @@ impl LedgerCleanupService {
-            blockstore.purge_slots(0, Some(lowest_slot));
-            *next_purge_batch += DEFAULT_PURGE_BATCH_SIZE;
-        }
+            debug!("purge: {} last_purge: {}", root, last_purge_slot);
+            let disk_utilization_pre = blockstore.storage_size();
 
-        let disk_utilization_post = blockstore.storage_size();
-
-        if let (Ok(disk_utilization_pre), Ok(disk_utilization_post)) =
-            (disk_utilization_pre, disk_utilization_post)
-        {
-            datapoint_debug!(
-                "ledger_disk_utilization",
-                ("disk_utilization_pre", disk_utilization_pre as i64, i64),
-                ("disk_utilization_post", disk_utilization_post as i64, i64),
-                (
-                    "disk_utilization_delta",
-                    (disk_utilization_pre as i64 - disk_utilization_post as i64),
-                    i64
-                )
+            // Notify blockstore of impending purge
+            //cleanup
+            let mut shreds = Vec::new();
+            let mut iterate_time = Measure::start("iterate_time");
+            let mut total_shreds = 0;
+            for (i, (slot, meta)) in blockstore.slot_meta_iterator(0).unwrap().enumerate() {
+                if i == 0 {
+                    debug!("purge: searching from slot: {}", slot);
+                }
+                total_shreds += meta.completed_data_indexes.len();
+                shreds.push((slot, meta.completed_data_indexes.len()));
+                if slot > root {
+                    break;
+                }
+            }
+            iterate_time.stop();
+            debug!(
+                "checking for ledger purge: {:?} max: {} slots: {} total_shreds: {} {}",
+                disk_utilization_pre,
+                max_ledger_utilization,
+                shreds.len(),
+                total_shreds,
+                iterate_time
             );
+            if (total_shreds as u64) < max_ledger_utilization {
+                return Ok(());
+            }
+            let mut cur_shreds = 0;
+            let mut lowest_slot_to_clean = shreds[0].0;
+            for (slot, num_shreds) in shreds.iter().rev() {
+                cur_shreds += *num_shreds as u64;
+                if cur_shreds > max_ledger_utilization {
+                    lowest_slot_to_clean = *slot;
+                    break;
+                }
+            }
+            if cur_shreds > 0 {
+                debug!(
+                    "cleaning up to: {} shreds: {}",
+                    lowest_slot_to_clean, total_shreds
+                );
+                let mut slot_update_time = Measure::start("slot_update");
+                *blockstore.lowest_cleanup_slot.write().unwrap() = lowest_slot_to_clean;
+                slot_update_time.stop();
+                let mut clean_time = Measure::start("ledger_clean");
+                blockstore.purge_slots(0, Some(lowest_slot_to_clean));
```

Comment:
Does it matter if this starts from 0 everytime? would it be more efficient to start from the last 
`lowest_slot_to_clean`?

Reply:
The old code made a limit of `DEFAULT_PURGE_BATCH_SIZE` number of slots deleted at a time so that the `WriteBatch` built by `purge_slots` wouldn't get too large. 

Here, the first time we hit `max_ledger_utilization` there will be a large number of deletes (not sure if there will be any issues), but then every subsequent delete should only be about 512 slots.

Reply:
Yea.. I actually tried individual deletes and ran into a lot of problems. Inserts got horribly slow and backed up. I was trying to delete non-root slots below root first to save space, but seems it's hard to get to work without deleterange.

Reply:
OK, maybe i can incorporate that batch size limit so WriteBatch is not too large.

Reply:
well with deleterange does it really count as increasing it? I see a pretty constant level performance, but I could do some more tests.

Reply:
@sakridge yeah I think with `DeleteRange` it's a non-issue :D. On the other hand, single deletes are just like inserting tombstones, which is probably why you saw that terrible insert perf.

Reply:
In my experiments, deleterange seems to take roughly constant time no matter the range size, so batching just increases the total time spent purging.